runtime/commonMain/src/kotlinx/serialization/ContextSerializer.kt

```diff
@@ -21,18 +20,28 @@ import kotlin.reflect.*
  * To create a serial module, use [SerializersModule] factory function.
  * To pass it to encoder and decoder, refer to particular [SerialFormat]'s documentation.
  */
-@ImplicitReflectionSerializer
-public class ContextSerializer<T : Any>(private val serializableClass: KClass<T>) : KSerializer<T> {
+@OptIn(UnsafeSerializationApi::class)
+public class ContextSerializer<T : Any>(
+    private val serializableClass: KClass<T>,
+    private val fallbackSerializer: KSerializer<T>?,
+    private val typeParametersSerializers: Array<KSerializer<*>>
+) : KSerializer<T> {
+
+    // Used from auto-generated code
+    public constructor(serializableClass: KClass<T>): this(serializableClass, null, EMPTY_ARRAY)
+
     public override val descriptor: SerialDescriptor =
         SerialDescriptor("kotlinx.serialization.ContextSerializer", UnionKind.CONTEXTUAL).withContext(serializableClass)
 
     public override fun serialize(encoder: Encoder, value: T) {
-        val serializer = encoder.context.getContextualOrDefault(value)
-        encoder.encodeSerializableValue(serializer, value)
+        val clz = value::class
+        val serializer = encoder.context.getContextual(clz) ?: clz.serializer()
+        @Suppress("UNCHECKED_CAST")
+        encoder.encodeSerializableValue(serializer as SerializationStrategy<T>, value)
     }
 
     public override fun deserialize(decoder: Decoder): T {
-        val serializer = decoder.context.getContextualOrDefault(serializableClass)
+        val serializer = decoder.context.getContextual(serializableClass) ?: serializableClass.serializerOrNull() ?: serializableClass.serializerNotRegistered()
```

Comment:
Same here, either use a fallbackSerializer or do not introduce it yet.

Reply:
Also, deserialize uses `serializerNotRegistered` while serialize uses `.serializer()` (although they throw the same exception)